File: 1981-minimize-the-difference-between-target-and-chosen-elements/1981-minimize-the-difference-between-target-and-chosen-elements.java
class Solution {
    public int minimizeTheDifference(int[][] mat, int target) {
        int m=mat.length,n=mat[0].length;
        Integer[][] dp=new Integer[m+1][10000];
        return rec(mat,0,m,n,target,0,dp);
    }
    int rec(int[][] mat,int idx,int m,int n,int target,int sum,Integer dp[][])
    {
        if(idx==m){
            return Math.abs(target-sum);
        }
        if(dp[idx][sum]!=null) return dp[idx][sum];
        
        int min=Integer.MAX_VALUE;
        for(int i=0;i<n;i++){
            min=Math.min(rec(mat,idx+1,m,n,target,mat[idx][i]+sum,dp),min);
        }
        return dp[idx][sum]=min;
    }
}

File: 1981-minimize-the-difference-between-target-and-chosen-elements/1981-minimize-the-difference-between-target-and-chosen-elements.cpp
class Solution {
public: int dp[71][56000];
    vector<vector<int>> a; int diff = INT_MAX ;
    
    int solve(int i,int sum,int k){
        if(i==a.size()){
            return  abs(sum-k);
        }
        if(dp[i][sum]!=-1) return dp[i][sum];
        
        int ans = INT_MAX;
        for(int j = 0;j<a[i].size();j++){
            ans = min(ans,solve(i+1,sum + a[i][j],k));
        }
        return dp[i][sum] =  ans;
    }
    
    int minimizeTheDifference(vector<vector<int>>& mat, int k) {
        memset(dp,-1,sizeof(dp));
        a = mat; return solve(0,0,k);
    }
};